orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py

```diff
@@ -13,150 +15,204 @@
+        self.materialize_grads: bool = False
+        self.materialize_grads_config: Optional[Tuple[torch.device, torch.dtype, torch.shape]] = None
+
+
+# Store the kernel specific information that cannot be retrieved and saved by PyTorch exporter.
+# For those infos that can only be retrieved with real run, we try to collect them in the first time run.
+# key: kernel_invoke_id, value: CustomFuncOpKernelInfo.
+_GlobalOpKernelInfoMap: Dict[str, CustomFuncOpKernelInfo] = {}
+
+
+def _get_context(forward_tensor_outputs: List[torch.Tensor]) -> Tuple[any, Optional[torch.Tensor]]:
+    """Search for context among all outputs.
+
+    Note1: All forward outputs of torch.autograd.Function shared the same gradient function pointer,
+        so here we just get the first tensor having grad_fn attribute.
+        (https://github.com/PyTorch/PyTorch/blob/15532595209d2daf34d35e10f8d3d3b64966aea2/torch/csrc/autograd/custom_function.cpp#L267)
+
+    Note2: Context can be None because NOT all torch.autograd.Function's are differentiable. The function
+        https://github.com/PyTorch/PyTorch/blob/d701357d921ef167d42c125e65b6f7da6be3ad0f/torch/csrc/autograd/custom_function.cpp#L209?
+        means if all output of forward function is not differentiable, then grad_fn will be None (not be set).
+
+        For example,
+            class Bar(torch.autograd.Function):
+                # A non-differentiable autograd Function whose forard output
+                # doesn't have grad_fn attribute.
+                @staticmethod
+                def forward(ctx, x):
+                    y = torch.ones_like(x)
+                    return y
+
+                @staticmethod
+                def backward(ctx, dy):
+                    dx = torch.zeros_like(dy)
+                    return dx
+
+    Returns:
+        ctx: context of the autograd.Function.
+        tensor: a tensor that owns the context.
+
     """
-    If the input is a DLPack tensor, we wrap it as a torch.Tensor and
-    set up its attributes according to other input flags. Otherwise,
-    we return the input as is.
-
-    grad_flag: indicate if "arg" requires gradient. This is only valid if
-            "arg" is a DLPack tensor.
-    tensor_flag: indicate if "arg" is a DLPack tensor.
-    inplace_flag: indicate if "arg" may be modified in custom function.
-    training_mode_flag: indicate if the top-level model is running
-                        under training (or inference) mode.
-    arg: a DLPack tensor or a normal Python object (e.g, a tuple of ints).
+    ctx = None
+    first_tensor_output = None
+    for arg in forward_tensor_outputs:
+        if not isinstance(arg, torch.Tensor) or not hasattr(arg, "grad_fn"):
+            continue
+        # Use the first context we see because all of arg's share the same one.
+        ctx = arg.grad_fn
+        first_tensor_output = arg
+        break
+
+    return (ctx, first_tensor_output)
+
+
+def _epilogize_forward(
```

Comment:
nit: 
```suggestion
def _finalize_forward(
```

Reply:
changed to _finalize_traing_mode_forward to better explain itself. :)